general/v0.1/registration.md

```diff
@@ -27,112 +27,96 @@ administrator. This MAY include but is not limited to the following:
+* `publicKey`: The public key of the FASP
 
 An example payload:
 
 ```
 {
-  "iss": "https://fasp.example.com",
-  "sub": "b2ks6vm8p23w",
-  "sec": "H4bo9pER2Ww4MlPs2Rf",
-  "exp": 1726498179
+  "name": "Example FASP",
+  "baseUrl": "https://fasp.example.com",
+  "clientId": "b2ks6vm8p23w",
+  "publicKey": "FbUJDVCftINc9FlgRu2jLagCVvOa7I2Myw8aidvkong=" 
 }
 ```
 
-FASPs SHOULD offer a way to re-generate the OAuth credentials and
-this token in case it gets lost or a fediverse server needs to be reinstalled
-from scratch.
+As a result the fediverse server MUST persist this information as a
+request for FASP registration and generate a client ID for the FASP and
+its own Ed25519 keypair for authenticating with the FASP. It MUST then
+reply with an HTTP status code `201` (Created) and a JSON object that
+contains the following keys and values:
 
-The following is a sketch of how this may look in the abstract:
-
-Step 1: A fediverse server admin is presented with a registration form
-
-![A bare-bones sign-up form asking for email, instance URL and acceptance of terms of service](../../images/instance_sign_up.svg)
-
-Step 2: Upon successful registration, the registration token is
-displayed
-
-![A webpage displaying a registration token with a button to copy to clipboard](../../images/instance_sign_up_success.svg)
+* `clientId`: The client ID
+* `publicKey`: The public key of the fediverse server
+* `redirectUri`: An URI to redirect to in order to finish the
+  registration
 
-## Adding a FASP to a Fediverse Server
-
-The option to add a new FASP SHOULD be part of the user interface
-for administrative settings that a fediverse software already has.
-
-To add a new FASP, the fediverse server adminstrator MUST enter the
-registration token obtained in the previous step. The fediverse
-software MUST validate the token before proceeding.
+An example payload:
 
-This validation MUST include:
+```json
+{
+  "clientId": "dfkl3msw6ps3",
+  "publicKey": "KvVQVgD4/WcdgbUDWH7EVaYX9W7Jz5fGWt+Wg8h+YvI=",
+  "redirectUri": "https://fedi.example.com/admin/fasps"
+}
+```
 
-* Ensuring the presence of the `iss`, `sub` and `sec` claims
-* Ensuring that `iss` is a valid HTTP(S) URI
-* If `exp` is present, ensure the date it represents is not in the past
+The FASP MUST persist this data and present the administrator with a
+page that explains how to finish the registration on their server.
 
-![A bare-bones web form to add a provider with a textarea for the registration token](../../images/add_provider.svg)
+To that end it MAY present a link to the `redirectUri` it received.
 
-After successful submission of this data, the fediverse software MUST
-persist it and initiate an OAuth 2.0 "client credentials" flow to
-obtain an access token. This token can then be used to authenticate
-subsequent API calls.
+It MUST display a fingerprint of the FASP's public key for comparison
+purposes. The fingerprint is the Base64 encoded SHA-256 hash of the
+public key.
```

Comment:
We might want to have something easier to compare than a base64 hash here, like an image or a sequence of words derived from the key? Does this need to even be derived from the key, or could it be part of the exchange which allows it to be arbitrary?

Reply:
Some short random number was my idea I originally mentioned in chat. We _could_ do that, but I went with this fingerprint instead for the following reasons:

* The security depends on the public key being the correct one. If someone were able to intercept the registration and manipulate it, they could leave a verification code intact and replace the public key. So by comparing fingerprints an administrator can not only be reasonably sure the registration is the one they just initiated, they can also be sure the most important part has not been tampered with.
* This is not for end-users. I think we can assume that administrators are familiar with the idea and process of comparing key fingerprints.

(BTW, I chose the format, SHA-256/Base64, because according to a quick google search that is what SSH uses for ed25519 keys).

Reply:
The problem with the "words" derived from the key is that it requires both parties to agree on a word-list, I think?